### ISSUE TITLE
Failure to perform outer-to-inner join conversion where the non-preserving side has a single-row aggregation

### ISSUE THREAD

**Issue by wangandi** [MaterializeInc/materialize#14684]
### What version of Materialize are you using?

latest main

### How did you install Materialize?

Built from source

### What is the issue?

This is an issue that is said to be the side-effect of #8223, but fixing #8223 did not actually fix the issue.`

### Relevant log output

```source
The new plan after #8223 was fixed:

materialize=> explain typed plan for select * from keyed_by_id_view left join lateral (select count(*) as matching from t1 where f1 = id) on true;
                      Optimized Plan                      
----------------------------------------------------------
 Source materialize.public.keyed_by_id_view (u2):        +
 | Project (#0..=#2)                                     +
                                                         +
 Source materialize.public.t1 (u4):                      +
 | Map dummy                                             +
 | Filter (#0) IS NOT NULL                               +
 | Project (#0, #2)                                      +
                                                         +
 Query:                                                  +
 %0 =                                                    +
 | Get materialize.public.keyed_by_id_view (u2)          +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ((#0))                                       +
 | Filter (#0) IS NOT NULL                               +
 | | types = (integer, bigint?, text?)                   +
 | | keys = ((#0))                                       +
 | Project (#0)                                          +
 | | types = (integer)                                   +
 | | keys = ((#0))                                       +
 | ArrangeBy (#0)                                        +
 | | types = (integer)                                   +
 | | keys = ((#0))                                       +
                                                         +
 %1 =                                                    +
 | Get materialize.public.t1 (u4)                        +
 | | types = (integer?, integer?)                        +
 | | keys = ()                                           +
 | Filter (#0) IS NOT NULL                               +
 | | types = (integer, integer?)                         +
 | | keys = ()                                           +
 | Project (#0)                                          +
 | | types = (integer)                                   +
 | | keys = ()                                           +
                                                         +
 %2 = Let l0 =                                           +
 | Join %0 %1 (= #0 #1)                                  +
 | | implementation = Differential %1 %0.(#0)            +
 | | types = (integer, integer)                          +
 | | keys = ()                                           +
 | Project (#0)                                          +
 | | types = (integer)                                   +
 | | keys = ()                                           +
 | Reduce group=(#0)                                     +
 | | agg count(true)                                     +
 | | types = (integer, bigint)                           +
 | | keys = ((#0))                                       +
                                                         +
 %3 =                                                    +
 | Get materialize.public.keyed_by_id_view (u2)          +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ((#0))                                       +
 | ArrangeBy (#0)                                        +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ((#0))                                       +
                                                         +
 %4 =                                                    +
 | Get %2 (l0)                                           +
 | | types = (integer, bigint)                           +
 | | keys = ((#0))                                       +
 | Project (#0)                                          +
 | | types = (integer)                                   +
 | | keys = ((#0))                                       +
 | Negate                                                +
 | | types = (integer)                                   +
 | | keys = ()                                           +
                                                         +
 %5 =                                                    +
 | Get materialize.public.keyed_by_id_view (u2)          +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ((#0))                                       +
 | Project (#0)                                          +
 | | types = (integer?)                                  +
 | | keys = ((#0))                                       +
                                                         +
 %6 =                                                    +
 | Union %4 %5                                           +
 | | types = (integer?)                                  +
 | | keys = ()                                           +
 | Map 0                                                 +
 | | types = (integer?, bigint)                          +
 | | keys = ()                                           +
                                                         +
 %7 =                                                    +
 | Union %2 %6                                           +
 | | types = (integer?, bigint)                          +
 | | keys = ((#0))                                       +
                                                         +
 %8 = Let l1 =                                           +
 | Join %3 %7 (= #0 #3)                                  +
 | | implementation = Differential %7 %3.(#0)            +
 | | types = (integer?, bigint?, text?, integer?, bigint)+
 | | keys = ((#0))                                       +
 | Project (#0..=#2, #4)                                 +
 | | types = (integer?, bigint?, text?, bigint)          +
 | | keys = ((#0))                                       +
                                                         +
 %9 =                                                    +
 | Get %8 (l1)                                           +
 | | types = (integer?, bigint?, text?, bigint)          +
 | | keys = ((#0))                                       +
 | Project (#0..=#2)                                     +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ((#0))                                       +
 | Negate                                                +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ()                                           +
                                                         +
 %10 =                                                   +
 | Get materialize.public.keyed_by_id_view (u2)          +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ((#0))                                       +
                                                         +
 %11 =                                                   +
 | Union %9 %10                                          +
 | | types = (integer?, bigint?, text?)                  +
 | | keys = ()                                           +
 | Map null                                              +
 | | types = (integer?, bigint?, text?, bigint?)         +
 | | keys = ()                                           +
                                                         +
 %12 =                                                   +
 | Union %8 %11                                          +
 | | types = (integer?, bigint?, text?, bigint?)         +
 | | keys = ((#0))                                       +
 
(1 row)
```
```
